src/v/model/transform.cc

```diff
@@ -83,10 +97,64 @@ operator<<(std::ostream& os, const transform_offset_options& opts) {
       },
       [&os](model::timestamp ts) {
           fmt::print(os, "{{ timequery: {} }}", ts.value());
+      },
+      [&os](model::transform_from_start off) {
+          fmt::print(os, "{{ offset: {} }}", off);
+      },
+      [&os](model::transform_from_end off) {
+          fmt::print(os, "{{ offset: {} }}", off);
       });
     return os;
 }
 
+/**
+ * Legacy version of transform_offset_options
+ *
+ * When writing out transform metadata, we write this version if and
+ * only if the position variant contains one of the legacy alternatives.
+ * This allows us to feature-gate the use of the new alternatives
+ * (see model/transform.h) without completely blocking transform deploys
+ * during an upgrade.
+ */
+struct legacy_transform_offset_options
+  : serde::envelope<
+      transform_offset_options,
```

Comment:
the first envelope template parameter looks wrong (CRTP)

Reply:
Good catch, thank you 🙏 
https://github.com/redpanda-data/redpanda/pull/20828